src/raw_rtde/connection.rs

```diff
@@ -0,0 +1,891 @@
+        let mut rdr = Cursor::new(payload);
+        let accepted = rdr
+            .read_u8()
+            .map_err(|_e| "Controller version reply too short")?;
+        Ok(accepted != 0)
+    }
+
+    fn encode_configure_inputs_request<U: AsRef<str>>(variables: &[U]) -> Result<Vec<u8>, String> {
+        let variables = variables.iter().map(|v| v.as_ref()).collect::<Vec<&str>>();
+        if !variables.iter().all(|v| !v.contains(',')) {
+            return Err(format!(
+                "Comma can't be part of variable name in {:?}",
+                variables
+            ));
+        }
+        let payload_as_string = variables.join(",");
+        let payload = payload_as_string.as_bytes().to_vec();
+        Self::encode_packet(RTDECommand::ConfigureInputs as u8, payload)
+    }
+
+    fn encode_configure_outputs_request<U: AsRef<str>>(
+        frequency: f64,
+        variables: &[U],
+    ) -> Result<Vec<u8>, String> {
+        let variables = variables.iter().map(|v| v.as_ref()).collect::<Vec<&str>>();
+        if !variables.iter().all(|v| !v.contains(',')) {
+            return Err(format!(
+                "Comma can't be part of variable name in {:?}",
+                variables
+            ));
+        }
+        let mut payload: Vec<u8> = vec![];
+        payload.write_f64::<BigEndian>(frequency).unwrap();
+        payload.extend(variables.join(",").as_bytes());
+
+        Self::encode_packet(RTDECommand::ConfigureOutputs as u8, payload)
+    }
+
+    fn map_type_name_to_discriminant(type_name: &str) -> Result<Discriminant<RTDEData>, String> {
+        let default_value = RTDEData::default_for_type_name(type_name)?;
+        Ok(discriminant(&default_value))
+    }
+
+    fn decode_configure_io_reply<U: AsRef<str>>(
+        payload: Vec<u8>,
+        variables: &[U],
+    ) -> Result<IoSetup, String> {
+        let variables = variables.iter().map(|v| v.as_ref()).collect::<Vec<&str>>();
+        let mut rdr = Cursor::new(payload);
+        let recipe_id: u8 = rdr.read_u8().map_err(|_e| "Inputs setup reply too short")?;
+        let payload = rdr.into_inner();
+        let string = String::from_utf8(payload[1..].to_vec())
+            .map_err(|_e| format!("Inputs reply contains non-UTF8 chars: {:?}", &payload[1..]))?;
+        let type_names: Vec<String> = string.split(',').map(String::from).collect();
+
+        if let Some(offender) = type_names.iter().position(|t| t == "NOT_FOUND") {
+            return Err(format!("Variable {} not found", variables[offender]));
+        }
+        if let Some(offender) = type_names.iter().position(|t| t == "IN_USE") {
+            return Err(format!(
+                "Variable {} is already in use by another connection",
+                variables[offender]
+            ));
+        }
+
+        let types: Vec<Discriminant<RTDEData>> = type_names
+            .iter()
+            .map(|n| Self::map_type_name_to_discriminant(n))
```

Comment:
Unnecessary closure, $\eta$-reduce it.

Reply:
This fails AFAIU due to the difference between `&String` and `&str`:
```
error[E0631]: type mismatch in function arguments
   --> ur_rtde/src/raw_rtde/connection.rs:273:18
    |
244 |     fn map_type_name_to_discriminant(type_name: &str) -> Result<Discriminant<RTDEData>, String> {
    |     ------------------------------------------------------------------------------------------- found signature of `for<'r> fn(&'r str) -> _`
...
273 |             .map(Self::map_type_name_to_discriminant)
    |              --- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected signature of `fn(&String) -> _`
    |              |
    |              required by a bound introduced by this call
    |
note: required by a bound in `map`
   --> /home/mikolajz/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:721:12
    |
721 |         F: FnMut(Self::Item) -> B,
    |            ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `map`
```

  Since `map_type_name_to_discriminant` is used only in this place, I can just change its definition to use `&String`, but my understanding is it makes it less generic (e.g., string literals are costly to convert to `&String`). What's the recommended way around such problems? Making `map_type_name_to_discriminant` a generic accepting both `&str` and `&String`?

Reply:
Ah, the joys of `Deref` coercions. Unfortunately closures are not coercion sites.

Well, in this case you could just keep the closure and add a comment as to why it is necessary.

Another alternative is to just make the function accept a `&String`, because it's a private function used in exactly one place, so you aren't forcing an allocating API on anyone with this change. It's probably not a great thing to do still, because you would then need *another* explanatory comment on the signature.

A third alternative is to insert a `.map(Deref::deref)` before the original `.map(Self::map_type_name_to_discriminant)` to have an actual iterator of `&str` items.

And yes, you could make the function generic over some type that's `AsRef<str>`.